### PR TITLE
Services/APT: Implemented PrepareToStartApplication and StartApplication

### PR DESCRIPTION
**Pull request by Subv**
This allows games to be launched from the Home Menu. ~However, there is still a bug with the GSP where the Home Menu doesn't release the GPU rights. It is unknown if the Home Menu should terminate itself after launching a new application.~ This was caused by incorrectly sending the Wakeup event as soon as the application called APT::Initialize, waiting until the application calls APT::Enable (as the real APT module does) fixes it by not starving the Home Menu threads and letting it finish its teardown sequence (although it still does not terminate, that's probably intentional)

To get the Home Menu to not hang when launching stuff, you need to have config block 0xF0006 (size 40 flags 8) in your config savegame, it doesn't matter if it's filled with zeros.

Home Menu must also be started with the LLE implementation of the NIM module, since we still have some bugs in the HLE NIM implementation that cause it to hang (specifically the 0x00110000 command).

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5207)

### CODE REVIEW

File: src/core/hle/service/apt/applet_manager.cpp [hunk 1]
```diff
@@ -539,6 +539,57 @@ ResultCode AppletManager::DoApplicationJump() {
...
+    // Note: APT first checks if we can launch the application via AM::CheckDemoLaunchRights and
+    // returns 0xc8a12403 if we can't. We intentionally do not implement that check.
+
+    // TODO(Subv): The APT service performs several checks here related to the exheader flags of the
+    // process we're launching and other things like title id blacklists. We do not yet implement
+    // any of that.
+
+    // Launch the title directly.
+    auto process =
+        NS::LaunchTitle(app_start_parameters.next_media_type, app_start_parameters.next_title_id);
```
Comment (B3n30) [inline src/core/hle/service/apt/applet_manager.cpp:574]: Are there checks if next_media_type and/or next_title_id aren't set?

File: src/core/hle/service/apt/apt.cpp [hunk 2]
```diff
@@ -461,39 +461,31 @@ void Module::APTInterface::GetProgramIdOnApplicationJump(Kernel::HLERequestConte
...
 void Module::APTInterface::StartApplication(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp(ctx, 0x1B, 3, 4); // 0x001B00C4
-    u32 buffer1_size = rp.Pop<u32>();
-    u32 buffer2_size = rp.Pop<u32>();
-    u32 flag = rp.Pop<u32>();
-    std::vector<u8> buffer1 = rp.PopStaticBuffer();
-    std::vector<u8> buffer2 = rp.PopStaticBuffer();
+    u32 parameter_size = rp.Pop<u32>();
+    u32 hmac_size = rp.Pop<u32>();
+    u8 paused = rp.Pop<u8>();
```
Comment (B3n30) [inline src/core/hle/service/apt/apt.cpp:480]: Is there any behavior different depending on this value? Is this really an u8 and not just a bool?

File: src/core/hle/service/apt/applet_manager.cpp [hunk 3]
```diff
@@ -539,6 +539,57 @@ ResultCode AppletManager::DoApplicationJump() {
...
+
+    // TODO(Subv): Send a WAKEUP signal via the apt parameter to the application. Right now our
+    // APT::Initialize implementation already does that because we have to support launching titles
+    // directly.
+
+    return RESULT_SUCCESS;
+}
+
+void AppletManager::SetDeliveryArg(std::vector<u8> parameter, std::vector<u8> hmac) {
+    auto& argument = delivery_arg.emplace();
```
Comment (B3n30) [inline src/core/hle/service/apt/applet_manager.cpp:588]: Is this used by the started application? maybe leave a TODO to implement the receiving side?

File: src/core/hle/service/apt/applet_manager.cpp (hunk 1 above)
Comment (Subv) [reply src/core/hle/service/apt/applet_manager.cpp:588]: There aren't any obvious ones, I think it just lets it fail when calling PM::LaunchTitle.

It's either that or it launches NATIVE_FIRM.

File: src/core/hle/service/apt/apt.cpp (hunk 2 above)
Comment (Subv) [reply src/core/hle/service/apt/apt.cpp]: It's a bool, if it's true then StartApplication will not send the Wakeup event, and the Home Menu has to use WakeupApplication to do it. I've implemented its behavior in AppletManager::StartApplication

File: src/core/hle/service/apt/applet_manager.cpp (hunk 3 above)
Comment (Subv) [reply src/core/hle/service/apt/applet_manager.cpp:647]: done

File: src/core/hle/service/apt/applet_manager.cpp (hunk 1 above)
Comment (Subv) [reply src/core/hle/service/apt/applet_manager.cpp]: I've added a few TODOs about this

**Comment by vvanelslande** [comment]
I can't start Pokémon Ultra Sun from the HOME Menu

**Comment by vvanelslande** [comment]
The HOME Menu only calls GetSystemTime

**Comment by Subv** [comment]
@vvanelslande Are you using LLE NIM and did you create the config block 0xF0006? Does the Menu freeze? A log would be nice

**Comment by vvanelslande** [comment]
> 
> 
> @vvanelslande Are you using LLE NIM and did you create the config block 0xF0006? Does the Menu freeze? A log would be nice

I'm using a build with this, #5196, and #5203 merged.

1. yes
2. yes
3. the screen says Nintendo 3DS
4. [log.txt](https://github.com/citra-emu/citra/files/4472824/log.txt)

**Comment by Subv** [comment]
The freeze could be caused by some Home Menu threads deadlocking due to there being no cross-core exclusive monitor support yet. Try forcing the system to use a single core instead of 2 and report back please

**Comment by vvanelslande** [comment]
@Subv that build doesn't have #4913 merged

File: src/core/hle/service/apt/applet_manager.cpp [hunk 4]
```diff
@@ -539,6 +552,103 @@ ResultCode AppletManager::DoApplicationJump() {
...
+        system.RequestShutdown();
+    }
+
+    app_start_parameters.reset();
+
+    if (!paused) {
+        // Send a Wakeup signal via the apt parameter to the application once it registers itself.
+        // The real APT service does this by spinwaiting on another thread until the application is
+        // registered.
+        MessageParameter wakeup_parameter{};
```
Comment (B3n30) [inline src/core/hle/service/apt/applet_manager.cpp:609]: Just call return(WakeupApplication()); here to avoid code duplication

**Comment by Subv** [comment]
@vvanelslande Could you try with the current master + #5203 + #5196 + this one, please?

**Comment by vvanelslande** [comment]
@Subv it doesn't boot.
I injected the game into Health and Safety Information.
The game didn't appear in the HOME Menu after I installed the CIA.

**Comment by Subv** [comment]
Do applets like the original Health & Safety app work?

**Comment by vvanelslande** [comment]
1. yes
2. latest
3. full system without [System Firmware](https://3dbrew.org/wiki/Title_list#00040138_-_System_Firmware)
4. [citra_log.txt](https://github.com/citra-emu/citra/files/4487263/citra_log.txt)

**Comment by B3n30** [comment]
The issue could be that the build valentin is using doesn't have core timing 2.0 which will result in the home menu tries to run the app on the same core as home menu which might lead to dead locks (similar to the games that got fixed to core timing 2.0). So as long as the reports come from such an outdated build, I don't think they are applicable here

**Comment by B3n30** [comment]
@Subv should we merge this?

**Comment by vvanelslande** [comment]
@B3n30 that build has the priority boost hack

**Comment by vvanelslande** [comment]
Pokémon Ultra Sun doesn't boot using Canary 1672, other apps boot.
[citra_log.txt](https://github.com/citra-emu/citra/files/4499458/citra_log.txt)

File: src/core/hle/service/apt/applet_manager.h [hunk 5]
```diff
@@ -167,10 +176,21 @@ class AppletManager : public std::enable_shared_from_this<AppletManager> {
...
+        FS::MediaType next_media_type;
+    };
+
 private:
     /// Parameter data to be returned in the next call to Glance/ReceiveParameter.
     std::optional<MessageParameter> next_parameter;
 
+    /// This parameter will be sent to the application/applet once they register themselves by using
+    /// APT::Initialize.
+    std::optional<MessageParameter> delayed_parameter;
```
Comment (B3n30) [inline src/core/hle/service/apt/applet_manager.h:190]: Do twe need to add code for handling save states with those new member variables?

**Comment by B3n30** [comment]
please fix conflicts

**Comment by FearlessTobi** [comment]
Superseded by https://github.com/citra-emu/citra/pull/6280.

**Pull request closed without merging**